design/application-abi.md

```diff
@@ -12,20 +12,25 @@ include.
 Current Unstable ABI
 --------------------
 
-The current WASI unstable ABI specifies only two exports from a WASI
-application:
+There are two kinds of modules:
 
-- `_start` - the program entry point
-- `memory` - linear memory used by the program
+ - A *command* exports a function named `_start`, with no arguments and no return
+   values. Environments shall call this function once, after instantiating the
+   module and all of its dependencies. After this function exits, the instance
+   is considered terminated and no further use of any of its exports should be
+   permitted.
 
-The `_start` export must be WebAssembly function and will be used as the program
-entry point.  This is the default name used by `lld` when linking WebAssembly
-modules.  The embedder is expected to call this function once the module is
-instantiated.
+ - A *reactor* exports a function named `_activate`, with no arguments and no
```

Comment:
Can you explain why you went with `_activate` rather then something more conventional like `_init` or `_initialize`?

Reply:
In conventional toolchains, `_init` is used in both commands and dynamic libraries. I was going for something that would connote that the presence of this export is what determines whether the thing is a command or an export.

Reply:
I think its reasonable to call this `_init`.   Its presence as a export can signal the application is a reactor.

Should we make these mutually exclusive too, such that exporting both these symbols is invalid?

Reply:
Good point about mutual exclusion. I've now added a sentence about that.

Concerning the name `_init`, I'm not attached to `_activate`, but I am nervous about having a thing named `_init` which is similar but subtly different from [the ELF `_init`](https://git.musl-libc.org/cgit/musl/tree/crt/crt1.c#n9) in several respects. The ELF `_init` isn't what gets called from the outside, it is used in executables as well as dynamic libraries, it's no longer used to run C++ constructors or libc init in most systems, and it can be awkwardly extended by end users through defining `.init` sections, and none of these apply to the `_activate` function I'm proposing.

Reply:
Still its is a better name IMO.    How about `_initialize` since it avoids that name clash and the needless contraction.

Reply:
Works for me. I've now updated the PR.